### DISCUSSION TITLE
Zephyr and MISRA C rule 14.4

### DISCUSSION THREAD

**Discussion by Laczen** [zephyrproject-rtos/zephyr#41604] [Q&A]
Hi, recently there was a discussion on MISRA C on the development mailing list. This triggered me to have a look at the required MISRA C rules for zephyr and one of the required rules is Rule 14.4:
`The controlling expression of an if statement and the controlling expression of an iteration-statement shall have essentially Boolean type`. 
Is my understanding of this rule correct that the following constructs are to be changed:
```
while (1) -> while (true)
uint32_t i: while (i) -> while (i != 0U)
int rc: if (!rc) -> if (rc == 0) and if (rc) -> if (rc != 0)
pointer *p: if (!p) -> if (p == NULL) and if (p) -> if (p != NULL)
```
and the following is no longer allowed:
```
for (uint32_t i=start; i--;)
```

If my understanding is correct it might be a good idea to communicate this requirement to all maintainers so that PR's are checked against this rule.

**Comment by yashi** [comment]
I assume we should.  But currently we enforce MISRA C under `kernel/`and base OS code (#9552).  I don't know what the "base OS code" means.

Just for a reference:

```
$ git grep -P 'for *\(.+[-+]+;' 
subsys/net/lib/sockets/sockets.c:       for (pfd = fds, i = nfds; i--; pfd++) {
subsys/net/lib/sockets/sockets.c:               for (pfd = fds, i = nfds; i--; pfd++) {
$ git describe
v2.7.99-2893-gc9a3dae6519
```
